pkgs/desktops/gnome-3/extensions/update-extensions.py

```diff
@@ -0,0 +1,245 @@
+    - Filter out versions that support old Gnome versions
+    - Download the extension and hash it
+    """
+    extension_versions: Dict[str, Dict[str, str]] = {}
+    for shell_version, version_prefix in supported_versions.items():
+        # Newest compatible extension version
+        extension_version: Optional[int] = max(
+            (
+                int(ext_ver)
+                for shell_ver, ext_ver in extension_version_map.items()
+                if (shell_ver.startswith(version_prefix))
+            ),
+            default=None,
+        )
+        # Extension is not compatible with this Gnome version
+        if not extension_version:
+            continue
+        logging.debug(
+            f"[{shell_version}] Downloading '{uuid}' v{str(extension_version)}"
+        )
+        sha256 = fetch_sha256sum(uuid, str(extension_version))
+        extension_versions[shell_version] = {
+            "version": str(extension_version),
+            "sha256": sha256,
+        }
+    return extension_versions
+
+
+def pname_from_url(url: str) -> Tuple[str, str]:
+    """
+    Parse something like "/extension/1475/battery-time/" and output ("battery-time", "1475")
+    """
+
+    url = url.split("/")  # type: ignore
+    return (url[3], url[2])
+
+
+def sanitize_string(string: str) -> str:
+    """
+    Replace thinkgs like "\\u0123" in strings. This is required for nix < 2.3.10, as it does not support "\\u"
+    """
+    # The encode-decode trick also produces "\\n", but Nix supports "\n".
+    return string.encode("unicode-escape").decode().replace("\\n", "\n")
+
+
+# Keep track of all names that have been used till now to detect collisions.
+# This works because we deterministically process all extensions in historical order
+# The outer dict level is the shell version, as we are tracking duplicates only per same Shell version.
+# key: shell version, value: Dict with key: pname, value: list of UUIDs with that pname
+package_name_registry: Dict[str, Dict[str, List[str]]] = {}
+for shell_version in supported_versions.keys():
+    package_name_registry[shell_version] = {}
+
+
+def process_extension(extension: Dict[str, Any]) -> Optional[Dict[str, Any]]:
+    """
+    Process an extension. It takes in raw scraped data and downloads all the necessary information that buildGnomeExtension.nix requires
+
+        Input: a json object of one extension queried from the site. It has the following schema (only important key listed):
+            {
+                "uuid": str,
+                "name": str,
+                "description": str,
+                "link": str,
+                "shell_version_map": {
+                    str: { "pk": int, "version": int },
+                    …
+                }
+            }
+
+            "uuid" is an extension UUID that roughly looks like this: "extension-name@username.domain.tld".
+            "link" follows the following schema: "/extension/$number/$extension-name/".
```

Comment:
Can you add a comment what pk is and use variables from the above dict in the link template?

Reply:
I instead went the opposite way and removed all unused keys, including `pk`. For the current packaging approach, it doesn't really matter anymore.

Reply:
I am still irked by $number and $extension-name, because I don‘t know what they are. It feels like they should be variables from the context but they aren‘t.

Reply:
I added a few lines to make it more clear that these are *not* existing variables from the context.